cpp/ql/lib/semmlecode.cpp.dbscheme

```diff
@@ -207,15 +207,62 @@ svnchurn(
  * [Locations](https://codeql.github.com/docs/writing-codeql-queries/providing-locations-in-codeql-queries/).
  */
 locations_default(
-    /** The location of an element that is not an expression or a statement. */
-    unique int id: @location_default,
-    int container: @container ref,
-    int startLine: int ref,
-    int startColumn: int ref,
-    int endLine: int ref,
-    int endColumn: int ref
+  /** The location of an element that is not an expression or a statement. */
+  unique int id: @location_default,
+  int container: @container ref,
+  int startLine: int ref,
+  int startColumn: int ref,
+  int endLine: int ref,
+  int endColumn: int ref
 );
 
+files(
+  unique int id: @file,
+  string name: string ref
+);
+
+folders(
+  unique int id: @folder,
+  string name: string ref
+);
+
+@container = @file | @folder
+
+containerparent(
+  int parent: @container ref,
+  unique int child: @container ref
+);
+
+/*- Lines of code -*/
+
+#keyset[element_id]
+numlines(
+  int element_id: @sourceline ref,
+  int num_lines: int ref,
+  int num_code: int ref,
+  int num_comment: int ref
+);
+
+/*- Diagnostic messages -*/
+
+diagnostics(
+  unique int id: @diagnostic,
+  int severity: int ref,
+  string error_tag: string ref,
+  string error_message: string ref,
+  string full_error_message: string ref,
+  int location: @location_default ref
+);
+
+case @diagnostic.severity of
+  10 = @diagnostic_debug
+| 20 = @diagnostic_info
+| 30 = @diagnostic_warning
+| 40 = @diagnostic_error
+;
```

Comment:
This simply does not hold for the trap produced by the c extractor, which produces severities in the range [0, ..., 5].

Reply:
Good point. This snippet comes from Ruby where we use this numbering scheme. In the end it does not matter too much which numbers are used. We could just as well adjust the Ruby extractor. If we can use the same numbers across languages that would be nice.